### PR TITLE
In app feedback

### PR DESCRIPTION
**Pull request by mbirghan**
This PR creates the backend code for sending the feedback to the guardian website. It includes:
- creation of the network request
- creation of the task to send the feedback message
- introducing a function that can be called to create and send the feedback

### CODE REVIEW

File: src/mozillavpn.cpp [hunk 1]
```diff
@@ -659,6 +660,25 @@ void MozillaVPN::removeDevice(const QString& deviceName) {
...
+  QString* buffer = new QString();
+  QTextStream* out = new QTextStream(buffer);
+
+  serializeLogs(out, [this, out, buffer, feedbackText, rating, category]{
+    Q_ASSERT(out);
+    Q_ASSERT(buffer);
+
+    delete out;
+    scheduleTask(new TaskSendFeedback(feedbackText, *buffer, rating, category));
+    delete buffer;
```
Comment (bakulf) [inline src/mozillavpn.cpp:675]: Why a task? You could send it directly, I guess. But, ignoring this part, please write a comment saying that the content of the "buffer" is copied by the TaskSendFeedback CTOR.

File: src/mozillavpn.cpp [hunk 2]
```diff
@@ -659,6 +660,25 @@ void MozillaVPN::removeDevice(const QString& deviceName) {
...
+  });
+
+
+  return;
```
Comment (bakulf) [inline src/mozillavpn.cpp:679]: remove this :)

File: src/networkmanager.cpp [hunk 3]
```diff
@@ -45,12 +62,7 @@ QByteArray NetworkManager::userAgent() {
 
     // System data
     out << "sys:";
-#ifdef MVPN_WASM
-    out << "WASM";
-#else
-    out << QSysInfo::productType().toLocal8Bit() << " "
-        << QSysInfo::productVersion().toLocal8Bit();
-#endif
+    out << NetworkManager::osVersion();
```
Comment (bakulf) [inline src/networkmanager.cpp:65]: `out << "sys:" << NetworkManager::osVersion();`

File: src/networkrequest.cpp [hunk 4]
```diff
@@ -294,6 +295,32 @@ NetworkRequest* NetworkRequest::createForHeartbeat(QObject* parent) {
   return r;
 }
 
+NetworkRequest* NetworkRequest::createForFeedback(QObject* parent, const QString& feedbackText, const QString& logs, const qint8 rating, const QString& category) {
+  NetworkRequest* r = new NetworkRequest(parent, 201);
+
+  // QUrl url(Constants::API_URL);
+  QUrl url("http://localhost:3000");
```
Comment (bakulf) [inline src/networkrequest.cpp:302]: :) no localhost hehe

File: src/networkrequest.cpp [hunk 5]
```diff
@@ -294,6 +295,32 @@ NetworkRequest* NetworkRequest::createForHeartbeat(QObject* parent) {
...
+  url.setPath("/api/v1/vpn/feedback");
+  r->m_request.setUrl(url);
+
+  r->m_request.setHeader(QNetworkRequest::ContentTypeHeader,
```
Comment (bakulf) [inline src/networkrequest.cpp:306]: What about the authentication header?

File: src/networkrequest.cpp [hunk 6]
```diff
@@ -294,6 +295,32 @@ NetworkRequest* NetworkRequest::createForHeartbeat(QObject* parent) {
...
+                         "application/json");
+
+  QJsonObject obj;
+  obj.insert("feedbackText", QJsonValue(feedbackText));
```
Comment (bakulf) [inline src/networkrequest.cpp:310]: I suspect you can drop the `QJsonValue` here.

File: src/tasks/sendfeedback/tasksendfeedback.cpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#include "models/user.h"
+#include "mozillavpn.h"
+#include "networkrequest.h"
+
+namespace {
+Logger logger(LOG_MAIN, "TaskSendFeedback");
+}
+
+TaskSendFeedback::TaskSendFeedback(const QString& feedbackText, const QString& logs, const qint8 rating, const QString& category)
+    : Task("TaskSendFeedback"), m_feedbackText(feedbackText.left(1000)), m_logs(logs.right(100000)), m_rating(rating), m_category(category) {
```
Comment (bakulf) [inline src/tasks/sendfeedback/tasksendfeedback.cpp:18]: Can we have:
```
constexpr uint32_t FEEDBACK_MESSAGE_MAX_LENGTH = 1000;
constexpr uint32_t FEEDBACK_LOG_MAX_LENGTH = 100000;
```

File: src/tasks/sendfeedback/tasksendfeedback.cpp [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+void TaskSendFeedback::run(MozillaVPN* vpn) {
+  logger.log() << "Sending the feedback";
+
+  NetworkRequest* request =
+      NetworkRequest::createForFeedback(this, m_feedbackText, m_logs, m_rating, m_category);
+
+  connect(request, &NetworkRequest::requestFailed,
+          [this, vpn](QNetworkReply::NetworkError error, const QByteArray&) {
+            logger.log() << "Failed to send feedback" << error;
+            vpn->errorHandle(ErrorHandler::toErrorType(error));
```
Comment (bakulf) [inline src/tasks/sendfeedback/tasksendfeedback.cpp:33]: No error handling here.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/mozillavpn.cpp (hunk 1 above)
Comment (mbirghan) [reply src/mozillavpn.cpp:675]: I saw that a lot of other neworkrequest are using tasks so I wanted to use the same pattern.

File: src/tasks/sendfeedback/tasksendfeedback.cpp (hunk 8 above)
Comment (mbirghan) [reply src/tasks/sendfeedback/tasksendfeedback.cpp:33]: The question is what type of error handling do we want to have?

Comment (mbirghan) [reply src/tasks/sendfeedback/tasksendfeedback.cpp:33]: I don't think we should display to the user that there was an error but we could try to resend it later. 
What do you think?

File: src/mozillavpn.cpp (hunk 1 above)
Comment (bakulf) [reply src/mozillavpn.cpp:675]: Sounds good to me. Just write a comment about the lifetime of buffer. Something like:
`TaskSendFeedback CTOR copies the content of buffer. We can delete it immediately  after.` Or something better.

File: src/tasks/sendfeedback/tasksendfeedback.cpp (hunk 8 above)
Comment (bakulf) [reply src/tasks/sendfeedback/tasksendfeedback.cpp:33]: ok! makes sense!

**Comment by mbirghan** [comment]
I added my idea for resending the feedback after it failed.

**Comment by bakulf** [comment]
I rebased this patch on top of the UI work. I close this PR but I keep all the original commits (squashed) and the authorship.

**Pull request closed without merging**